### PR TITLE
Support warm starting in backtest()

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2311 
Pair coded with @martinjankowiak 

## Tested
- [x] smoke test with many window shapes

### CODE REVIEW

File: pyro/contrib/forecast/evaluate.py [hunk 1]
```diff
@@ -173,3 +189,38 @@ def backtest(data, covariates, model_fn, guide_fn=None, *,
         del pred
 
     return results
+
+
+def _warm_start_param_store(t0, t1, t0_old, t1_old):
```
Comment (fritzo) [inline pyro/contrib/forecast/evaluate.py]: @fehiepsi can you please review the indexing logic in this function?

**Comment by fritzo** [comment]
@martinjankowiak please let me know if you need any additional functionality.
💯if you get a chance to test on your problem before merging 🙂

File: pyro/contrib/forecast/evaluate.py [hunk 2]
```diff
@@ -173,3 +189,38 @@ def backtest(data, covariates, model_fn, guide_fn=None, *,
...
+        if time_dim is None:
+            continue
+        assert time_dim < 0
+        time_dim += constrained_param.dim() - param.dim()
+        with torch.no_grad():
+
+            # Truncate params that have just left the window.
+            if t0 != t0_old:
+                dots = (slice(None),) * (param.dim() + time_dim)
+                param = param[dots + (slice(t0_old - t0, None),)]
```
Comment (fehiepsi) [inline pyro/contrib/forecast/evaluate.py]: I think `t0_old - t0` should be `t0 - t0_old`. In addition, you might not need the condition `if t0 != t0_old`.

**Comment by fritzo** [comment]
Closing as this will not handle `LinearHMM` auxiliary parameters because they occur outside of the time plate. We'll need something closer to `PrefixReplayMessenger` logic...

**Pull request closed without merging**